dkulib/parallelizer/parallelizer.py

```diff
@@ -0,0 +1,298 @@
+        batch_response_parser: Callable[[List[Dict], Any, NamedTuple], List[Dict]] = DEFAULT_RESPONSE_PARSER,
+        output_column_prefix: AnyStr = DEFAULT_OUTPUT_COLUMN_PREFIX,
+        verbose: bool = DEFAULT_VERBOSE,
+    ):
+        self.function = function
+        self.error_handling = error_handling
+        self.exceptions_to_catch = exceptions_to_catch
+        if error_handling == ErrorHandling.LOG and not exceptions_to_catch:
+            raise ValueError(
+                "Please set at least one exception in exceptions_to_catch")
+        self.parallel_workers = parallel_workers
+        self.batch_support = batch_support
+        if not batch_support:
+            # Overwrite necessary args for row-by-row iteration
+            batch_size = 1
+            batch_response_parser = _parse_batch_response_default
+            
+        self.batch_size = batch_size
+        self.batch_response_parser = batch_response_parser
+        self.output_column_prefix = output_column_prefix
+        self.verbose = verbose
+        self._output_column_names = None  # Will be set at runtime by the run method
+
+    def _get_unique_output_column_names(self, existing_names: List[AnyStr]) -> NamedTuple:
+        """Returns a named tuple with prefixed column names and their descriptions"""
+        OutputColumnNameTuple = namedtuple(
+            "OutputColumnNameTuple", self.OUTPUT_COLUMN_NAME_DESCRIPTIONS.keys())
+        return OutputColumnNameTuple(
+            *[
+                generate_unique(
+                    name=column_name, existing_names=existing_names, prefix=self.output_column_prefix)
+                for column_name in OutputColumnNameTuple._fields
+            ]
+        )
+
+    def _apply_function_and_parse_response(
+        self, batch: List[Dict] = None, **function_kwargs,
+    ) -> Union[Dict, List[Dict]]:  # sourcery skip: or-if-exp-identity
+        """Wraps a row-by-row or batch function with error logging and response parsing
+        It applies `self.function` and:
+        - If batch, parse the function response to extract results and errors using `self.batch_response_parser`
+        - handles errors from the function with two methods:
+            * (default) log the error message as a warning and return the row with error keys
+            * fail if there is an error (if `self.error_handling == ErrorHandling.FAIL`)
+        """
+        output = deepcopy(batch)
+        for output_column in self._output_column_names:
+            for output_row in output:
+                output_row[output_column] = ""
+        try:
+            if not self.batch_support:
+                # In the row-by-row case, there is only one element in the list as batch_size=1
+                response = (
+                    self.function(row=batch[0], **function_kwargs)
+                )
+            else:
+                response = (
+                    self.function(batch=batch, **function_kwargs)
+                )
```

Comment:
```suggestion

              # In the row-by-row case, there is only one element in the list as batch_size=1
              response = (
                  self.function(row=batch[0:batch_size-1], **function_kwargs)
              )
           
```

Reply:
hahah this one if/else is nagging me as well, but i think the proposed solution does not work, because Python always returns a list when you use slicing

Reply:
You are right, my bad. Maybe we can always work on batches and remove the row parameter. If people want to operate row by row the will create batches with one row.

Reply:
Maybe we can just forget about that comment as it would change the spec quite a lot just for a refactor

Reply:
Agreed that the spec should not change: if the user works with a row-by-row API (majority of cases) the input/output should be row.